### ISSUE TITLE
compiler panic `cat_expr_unadjusted Errd` (used to work)

### ISSUE THREAD

**Issue by wagnerf42** [rust-lang/rust#41498]
hi, my code used to compile last week.
I updated the compiler today and now it crashes
note that you can get my code here (https://github.com/wagnerf42/Jimn/tree/master/rust/jimn)

```
~/c/J/r/jimn ❯❯❯ rustup update                                                                                       ⏎
info: syncing channel updates for 'nightly-x86_64-unknown-linux-gnu'

  nightly-x86_64-unknown-linux-gnu unchanged - rustc 1.18.0-nightly (2bd4b5c6d 2017-04-23)

~/c/J/r/jimn ❯❯❯ export RUST_BACKTRACE=1

~/c/J/r/jimn ❯❯❯ cargo run --example bentley_ottmann ./tests_bentley_ottmann/overlap.bo
   Compiling jimn v0.1.0 (file:///home/wagnerf/code/Jimn/rust/jimn)
error: internal compiler error: cat_expr_unadjusted Errd
   --> src/bentley_ottmann.rs:165:9
    |
165 | /         self.events_data
166 | |             .entry(event_point)
167 | |             .or_insert_with(|| {
168 | |                                 events.push(event_point);
169 | |                                 [HashSet::new(), HashSet::new()]
170 | |                             })
171 | |             [event_type]
    | |________________________^

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: run with `RUST_BACKTRACE=1` for a backtrace

thread 'rustc' panicked at 'Box<Any>', /checkout/src/librustc_errors/lib.rs:375
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
             at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at /checkout/src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at /checkout/src/libstd/sys_common/backtrace.rs:60
             at /checkout/src/libstd/panicking.rs:355
   3: std::panicking::default_hook
             at /checkout/src/libstd/panicking.rs:365
   4: std::panicking::rust_panic_with_hook
             at /checkout/src/libstd/panicking.rs:549
   5: std::panicking::begin_panic
   6: rustc_errors::Handler::abort_if_errors
   7: rustc_passes::consts::check_crate
   8: rustc_driver::driver::phase_3_run_analysis_passes::{{closure}}
   9: rustc::ty::context::TyCtxt::create_and_enter
  10: rustc_driver::driver::phase_3_run_analysis_passes
  11: rustc_driver::driver::compile_input
  12: rustc_driver::run_compiler
  13: std::panicking::try::do_call
  14: __rust_maybe_catch_panic
             at /checkout/src/libpanic_unwind/lib.rs:98
  15: <F as alloc::boxed::FnBox<A>>::call_box
  16: std::sys::imp::thread::Thread::new::thread_start
             at /checkout/src/liballoc/boxed.rs:650
             at /checkout/src/libstd/sys_common/thread.rs:21
             at /checkout/src/libstd/sys/unix/thread.rs:84
  17: start_thread
  18: clone

error: Could not compile `jimn`.
```

**Comment by wagnerf42** [comment]
note: going back to nightly-2017-04-01; everything work again.  'nightly-2017-04-17 is not ok. so troubles started somewhere in between.

**Comment by eddyb** [comment]
cc @nikomatsakis

**Comment by wagnerf42** [comment]
not too sure what to do with this message.
can I help you with something ? I can do a small dichotomy to figure out the last working nightly if this can help.

**Comment by arielb1** [comment]
Minified:
```Rust
struct S;
impl S {
    fn mutate(&mut self) {}
}

fn call_and_ref<T, F: FnOnce() -> T>(x: &mut Option<T>, f: F) -> &mut T {
    *x = Some(f());
    x.as_mut().unwrap()
}

fn main() {
    let mut n = None;
    call_and_ref(&mut n, || [S])[0].mutate();
}
```

**Comment by nikomatsakis** [comment]
@wagnerf42 there is a pending fix (#41578), no action needed on your part

**Comment by nikomatsakis** [comment]
triage: P-high
